openwisp_radius/api/views.py

```diff
@@ -290,75 +278,50 @@ class AccountingView(TokenAuthorizationMixin, ListCreateAPIView):
+        # Create or Update
+        try:
+            instance = self.get_queryset().get(unique_id=data.get('unique_id'))
+        except RadiusAccounting.DoesNotExist:
+            serializer = self.get_serializer(data=data)
+            serializer.is_valid(raise_exception=True)
+            valid_data = self._change_valid_data(serializer.validated_data.copy())
+            serializer.create(valid_data)
             headers = self.get_success_headers(serializer.data)
             return Response(None, status=201, headers=headers)
-        # trying to create a record which
-        # already exist, fallback to update
-        if errors == 1 and 'unique_id' in error_keys:
-            return self.update(request, *args, **kwargs)
         else:
-            raise RestValidationError(serializer.errors)
-
-    def perform_create(self, serializer):
-        if app_settings.API_ACCOUNTING_AUTO_GROUP:
-            user_model = get_user_model()
-            username = serializer.validated_data.get('username', '')
-            try:
-                user = user_model.objects.get(username=username)
-            except User.DoesNotExist:
-                logging.info(f'no corresponding user found for username: {username}')
-                serializer.save()
-            else:
-                group = user.radiususergroup_set.order_by('priority').first()
-                # user may not have a group defined
-                groupname = group.groupname if group else None
-                serializer.save(groupname=groupname)
-        else:
-            return super().perform_create(serializer)
-
-    def update(self, request, *args, **kwargs):
-        try:
-            instance = self.get_queryset().get(unique_id=request.data['unique_id'])
-        # trying to update a record which
-        # does not exist, fallback to create
-        except RadiusAccounting.DoesNotExist:
-            return self.create(request, *args, **kwargs)
-        serializer = self.get_serializer(instance, data=request.data, partial=False)
-        serializer.is_valid(raise_exception=True)
-        serializer.save()
-        return Response(None)
+            serializer = self.get_serializer(instance, data=data, partial=False)
+            serializer.is_valid(raise_exception=True)
+            valid_data = self._change_valid_data(serializer.validated_data.copy())
+            serializer.update(instance, valid_data)
+            return Response(None)
 
-    def _get_status_type(self, request):
+    def _pre_validation(self, invalid_data):
         try:
-            return request.data['status_type']
+            status_type = invalid_data['status_type']
         except KeyError:
             raise RestValidationError({'status_type': [_('This field is required.')]})
+        for field in ['session_time', 'input_octets', 'output_octets']:
+            if status_type == 'Start' and invalid_data[field] == '':
+                invalid_data[field] = 0
```

Comment:
I think we could do this in the serializer

Reply:
I'm not sure about two things:

- why is the argument called `invalid_data`? Is not that the data is invalid, it's our API which is a bit different and needs to preprocess the data in order to [be more forgiving with what it receives](https://en.wikipedia.org/wiki/Robustness_principle)
- have you already tried to override the `validate` method of the serializer to perform these operations there?

Something like:

```python
def validate(self, ...)
    # do these operations before anything
    return super().validate(.........)
```

What is it that you tried exactly?

Reply:
PS: alternatively, we may also use [field-level validation](https://www.django-rest-framework.org/api-guide/serializers/#field-level-validation)